[view-transitions] Back icon jiggles while animating on https://codepen.io/bramus/full/mdowgYX.

mattwoodrow opened the pull request:
#### e6a8ede67b0919c59b9bef1c1368c8869c1f14b7
<pre>
[view-transitions] Back icon jiggles while animating on <a href="https://codepen.io/bramus/full/mdowgYX.">https://codepen.io/bramus/full/mdowgYX.</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=274606">https://bugs.webkit.org/show_bug.cgi?id=274606</a>
&lt;<a href="https://rdar.apple.com/128627261">rdar://128627261</a>&gt;

Reviewed by Antoine Quint.

If the view transition is capturing the root element, then we reparent the GraphicsLayer for
the RenderView instead. Make sure we request a geometry update on the right RenderLayer to
account for this.

New non-WPT test added, since this requires overlay scrollbars to be enabled to reproduce (otherwise
the scrollbar move requests a geometry update on the RenderView as well and hides the bug).

This also exposed a bug where the transition from &apos;active&apos; to &apos;after&apos; state of the view transition
animations wasn&apos;t invalidating layer composition, but does affect compositing decisions.

The original testcase still isn&apos;t perfect with these changes, since async scrolling still
incorrectly accounts for the view transition overlays. That will be fixed by bug 273612.

* LayoutTests/fast/css/view-transitions-scrolling-position-fixed-expected.html: Added.
* LayoutTests/fast/css/view-transitions-scrolling-position-fixed.html: Added.
* Source/WebCore/animation/CSSTransition.cpp:
(WebCore::CSSTransition::resolve):
* Source/WebCore/animation/CSSTransition.h:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::apply):
* Source/WebCore/animation/KeyframeEffect.h:
* Source/WebCore/animation/KeyframeEffectStack.cpp:
(WebCore::KeyframeEffectStack::applyKeyframeEffects):
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::resolve):
* Source/WebCore/animation/WebAnimation.h:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::updatePseudoElementStyles):

Canonical link: <a href="https://commits.webkit.org/279674@main">https://commits.webkit.org/279674@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6e0ba3d0874fd8173f941539d159bc6f07c15a6a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53958 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33330 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6489 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57236 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4683 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56261 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40837 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4569 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43690 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3088 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56055 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31537 "Found 3 new test failures: animations/3d/change-transform-in-end-event.html, compositing/culling/clear-fixed-iframe.html, compositing/overflow/nested-scrolling.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46678 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24831 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28376 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4000 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2834 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/50021 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4203 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58829 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29135 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/4281 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51105 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30325 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46815 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50440 "3 api tests failed or timed out") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11797 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/31276 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30100 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/fast/canvas/view-transitions-scrolling-position-fixed-expected.html:
fast/css might be better?

mattwoodrow replied:
I... guess I used tab-complete for this? It definitely should be in `css`, not `canvas`.

graouts commented on Source/WebCore/animation/KeyframeEffect.cpp:
```diff
@@ -1316,7 +1316,11 @@ void KeyframeEffect::apply(RenderStyle& targetStyle, const Style::ResolutionCont
+        if (m_phaseAtLastApplication != computedTiming.phase) {
+            m_phaseAtLastApplication = computedTiming.phase;
+            m_target->invalidateStyleAndLayerComposition();
+        }
```
I'm not sure this is the right time to call this. You'll note that we otherwise call `invalidateStyleForAnimation()` at moments when timing properties change, but the `apply()` method is already likely called after the target was marked for invalidation and style resolution is in progress. Cc'ing @anttijk who might have some thoughts on the topic.

mattwoodrow replied:
That's a good point.

`invalidateStyleForAnimation` doesn't actually work here, since the transition from 'active' to 'after' might not actually have a style difference (and frequently seems to not with view transitions).

`invalidateStyleForAnimation` forces the resulting restyle to have `StyleDifference::RecompositeLayer`, even if the computed style is otherwise identical.

`RenderLayerCompositor` makes different layerization decisions based on phase (using `isCurrentlyAffectingProperty`), so it also needs to know when the result of this will change.

I guess we ned to either to either find the right spot to replace a `invalidateStyleForAnimation` call with `invalidateStyleAndLayerComposition`, or alternatively just invalidate the layer composition state more directly (which feels like a bit of a layering violation here).

mattwoodrow replied:
Though about this more, and I don't think we can compute it earlier, since it depends on the `startTime` passed to apply.

The caller of `apply` currently computes an `AnimationImpact`, so I've refactored it a bit so that we can include this required composition into that.

graouts reviewed (changes requested):
Code change looks good, but please check on the red bots.

mattwoodrow committed:
[view-transitions] Back icon jiggles while animating on https://codepen.io/bramus/full/mdowgYX.
https://bugs.webkit.org/show_bug.cgi?id=274606
<rdar://128627261>

Reviewed by Antoine Quint.

If the view transition is capturing the root element, then we reparent the GraphicsLayer for
the RenderView instead. Make sure we request a geometry update on the right RenderLayer to
account for this.

New non-WPT test added, since this requires overlay scrollbars to be enabled to reproduce (otherwise
the scrollbar move requests a geometry update on the RenderView as well and hides the bug).

This also exposed a bug where the transition from 'active' to 'after' state of the view transition
animations wasn't invalidating layer composition, but does affect compositing decisions.

The original testcase still isn't perfect with these changes, since async scrolling still
incorrectly accounts for the view transition overlays. That will be fixed by bug 273612.

* LayoutTests/fast/css/view-transitions-scrolling-position-fixed-expected.html: Added.
* LayoutTests/fast/css/view-transitions-scrolling-position-fixed.html: Added.
* Source/WebCore/animation/CSSTransition.cpp:
(WebCore::CSSTransition::resolve):
* Source/WebCore/animation/CSSTransition.h:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::apply):
* Source/WebCore/animation/KeyframeEffect.h:
* Source/WebCore/animation/KeyframeEffectStack.cpp:
(WebCore::KeyframeEffectStack::applyKeyframeEffects):
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::resolve):
* Source/WebCore/animation/WebAnimation.h:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::updatePseudoElementStyles):

Canonical link: https://commits.webkit.org/279674@main
